File: backend/app/schemas/study.py
# backend/app/schemas/study.py
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


# One stop in the participant flow, as exposed to the client.
class StepPublic(BaseModel):
    id: str                       # e.g. "quiz:base", "survey:post_links"
    kind: Literal["quiz", "survey"]
    key: str                      # quiz_id for quizzes, response stage for surveys
    label: str
    route: str                    # frontend path for this step
    variant: Optional[str] = None
    completed: bool = False


# Returned by GET /study/next - the single authority the frontend uses to decide
# where a participant goes. Replaces the routing ladders that used to be
# duplicated across the quiz page, the survey page, and both backend services.
class NextStepResponse(BaseModel):
    next_step: Optional[StepPublic] = None
    next_route: str               # where to send the user right now
    completed_count: int = 0
    total_steps: int = 0
    finished: bool = False


# Returned by GET /study/flow - the participant's whole assigned journey, used
# for progress display ("Step 3 of 9").
class StudyFlowResponse(BaseModel):
    steps: List[StepPublic]
    variant_sequence: List[str] = Field(default_factory=list)
    current_step_id: Optional[str] = None
    completed_count: int = 0
    total_steps: int = 0
    finished: bool = False
    study_flow_version: int = 1
    mode: str = "all_variants"


# Admin view of the active flow configuration.
class StudyConfigPublic(BaseModel):
    mode: Literal["all_variants", "single_variant"]
    variant_order: List[str]
    counterbalance: bool
    version: int
    # Every variant the code currently declares, so the admin UI can render the
    # full list even if the stored order is stale.
    known_variants: List[str] = Field(default_factory=list)
    variant_labels: dict[str, str] = Field(default_factory=dict)
    # Worked example of the orders participants will actually receive.
    preview: List[List[str]] = Field(default_factory=list)


# Admin payload for changing the flow. All fields optional - partial update.
class StudyConfigUpdate(BaseModel):
    mode: Optional[Literal["all_variants", "single_variant"]] = None
    variant_order: Optional[List[str]] = None
    counterbalance: Optional[bool] = None

File: frontend/lib/study.ts
// frontend/lib/study.ts
//
// The study flow is resolved by the backend. Pages ask "where do I go next?"
// rather than each re-deriving the answer from the user's progress flags — the
// routing ladders that used to live in quiz/[quiz_id].tsx and survey.tsx are
// now a single call to /study/next.
import { apiFetch } from "./fetcher";

export type StepKind = "quiz" | "survey";

export type StudyStep = {
  id: string;        // e.g. "quiz:base", "survey:post_links"
  kind: StepKind;
  key: string;       // quiz_id for quizzes, survey stage for surveys
  label: string;
  route: string;     // where this step lives
  variant?: string | null;
  completed: boolean;
};

export type NextStepResponse = {
  next_step: StudyStep | null;
  next_route: string;
  completed_count: number;
  total_steps: number;
  finished: boolean;
};

export type StudyFlowResponse = {
  steps: StudyStep[];
  variant_sequence: string[];
  current_step_id: string | null;
  completed_count: number;
  total_steps: number;
  finished: boolean;
  study_flow_version: number;
  mode: string;
};

export type StudyConfig = {
  mode: "all_variants" | "single_variant";
  variant_order: string[];
  counterbalance: boolean;
  version: number;
  known_variants: string[];
  variant_labels: Record<string, string>;
  preview: string[][];
};

export type StudyConfigUpdate = {
  mode?: "all_variants" | "single_variant";
  variant_order?: string[];
  counterbalance?: boolean;
};

export async function getNextStep() {
  return apiFetch<NextStepResponse>("/api/study/next");
}

export async function getStudyFlow() {
  return apiFetch<StudyFlowResponse>("/api/study/flow");
}

// --- admin ---

export async function getStudyConfig() {
  return apiFetch<StudyConfig>("/api/study/config");
}

export async function updateStudyConfig(patch: StudyConfigUpdate) {
  return apiFetch<StudyConfig>("/api/study/config", {
    method: "PUT",
    body: JSON.stringify(patch),
  });
}

/** Step id for a quiz — mirrors study_flow.quiz_step_id on the backend. */
export function quizStepId(quizId: string): string {
  return `quiz:${quizId}`;
}

/** Step id for a survey stage — mirrors study_flow.survey_step_id. */
export function surveyStepId(stage: string): string {
  return `survey:${stage}`;
}
